android/src/main/java/com/amplitude/android/migration/RemnantDataMigration.kt

```diff
@@ -0,0 +1,177 @@
+package com.amplitude.android.migration
+
+import com.amplitude.common.android.LogcatLogger
+import com.amplitude.core.Amplitude
+import com.amplitude.core.Storage
+import com.amplitude.core.utilities.optionalJSONObject
+import org.json.JSONObject
+
+/**
+ * When switching the SDK from previous version to this version, remnant data might remain unsent in sqlite.
+ * This migration:
+ *      1. reads device/user id, events, identifies from sqlite tables
+ *      2. converts the events and identifies to JsonObjects
+ *      3. saves the device/user id, converted events and identifies to current storage
+ *      4. deletes data from sqlite table
+ */
+
+class RemnantDataMigration(
+    val amplitude: Amplitude,
+) {
+    companion object {
+        const val DEVICE_ID_KEY = "device_id"
+        const val USER_ID_KEY = "user_id"
+        const val LAST_EVENT_TIME_KEY = "last_event_time"
+        const val LAST_EVENT_ID_KEY = "last_event_id"
+        const val PREVIOUS_SESSION_ID_KEY = "previous_session_id"
+    }
+
+    lateinit var databaseStorage: DatabaseStorage
+
+    suspend fun execute() {
+        databaseStorage = DatabaseStorageProvider.getStorage(amplitude)
+
+        val firstRunSinceUpgrade = amplitude.storage.read(Storage.Constants.LAST_EVENT_TIME)?.toLongOrNull() == null
+
+        moveDeviceAndUserId()
```

Comment:
If any of these `move` method fails, this continues running, as the `catch` block will catch any exception in each method. As there is no "transaction" here, will this cause any data inconsistency issue. I think it is probably fine to leave like this, just to double check.

Reply:
The `move` methods are independent of each other.

`moveDeviceAndUserId`/ `moveSessionData` are safe - they overwrite current value only if the current value is not set.
`moveInterceptedIdentifies`/ `moveIdentifies` run once only - an event are skipped on exception
`moveEvents` can run several times for en event (say, the event is written to new storage but removing from legacy storage fails for some reason). `insert_id` should prevent event duplication in this case

It looks like legacy field name is `uuid`, new name is `insert_id` - I've added conversion `uuid` -> `insert_id`